workflow/restore_from_mnemonic: restrict choices for the last word

benma opened the pull request:
When one has entered 23 words, the valid choices for the
last word can be restricted to a subset of all bip39 words for which
the seedphrase checksum is valid.

This makes the entry of the last word a lot quicker 8 choices instead
of 2048.

For 12/18 words, there are a lot more valid words, so we cannot show
them in a menu. We could restrict keyboard entry to these choices, but
it has a big UX downside: if there is a typo in the words before,
you'd get stuck in the last word, as your word is not present, and
there would be no message about invalid recovery words. This is why we
only enable this feature for 24 words. In this case, all choices can
be shown in a menu with an additional entry 'None of them'.

This also enables an expert user to manually create and enter a seed
using e.g. dice throws, without using other software to compute the
checksum. They can pick an arbitrary word instead. This leads to an
entropy reduction from 256 bits to 253 bits for 24 words, which is not
a problem.

My1 commented:
> it has a big UX downside: if there is a typo in the words before,
> you'd get stuck in the last word, as your word is not present, and
> there would be no message about invalid recovery words.

that's my main worry.

> They can pick an arbitrary word instead. This leads to an
> entropy reduction from 256 bits to 253 bits for 24 words, which is not
> a problem.

there are dices with 8 sides so even if you have that idea of building words yourself, you can work around the entropy loss.

the Idea is pretty cool tho.

benma commented:
> that's my main worry.

it's been solved btw. we only allow choosing the last word for 24 words, not for 12/18. No such ux downside there.

My1 commented:
> > that's my main worry.
> 
> it's been solved btw. we only allow choosing the last word for 24 words, not for 12/18. No such ux downside there.

How is Going choice only on 24 words do anything about the stuck because typo issue?

benma commented:
Rebased on https://github.com/digitalbitbox/bitbox02-firmware/pull/687 - ported to Rust - so that the code with all the menus is easier to follow.

@My1 I updated the commit message to explain:

> For 12/18 words, there are a lot more valid words, so we cannot show
them in a menu. We could restrict keyboard entry to these choices, but
it has a big UX downside: if there is a typo in the words before,
you'd get stuck in the last word, as your word is not present, and
there would be no message about invalid recovery words. This is why we
only enable this feature for 24 words. In this case, all choices can
be shown in a menu with an additional entry 'None of them'.

benma commented:
This is ready now @x1ddos

x1ddos reviewed (approved):
LGTM

benma commented on src/rust/bitbox02-rust/src/workflow/mnemonic.rs:
```diff
@@ -59,6 +66,55 @@ pub async fn confirm_word(choices: &[&str], title: &str) -> Result<u8, CancelErr
+            // Set last three bits of the seed to `i`.
+            seed[31] &= 0b11111000;
+            seed[31] |= i;
+            // Computer checksum.
```
Should have been Compute checksum :joy:
